src/librustc_target/abi/mod.rs

```diff
@@ -102,8 +118,16 @@ impl TargetDataLayout {
                 [p @ "p", s, ref a..] | [p @ "p0", s, ref a..] => {
                     dl.pointer_size = size(s, p)?;
                     dl.pointer_align = align(a, p)?;
-                }
-                [s, ref a..] if s.starts_with("i") => {
+                    resize_and_set(&mut dl.pointers, 0, Some((dl.pointer_size,
+                                                              dl.pointer_align)));
+                },
+                [p, s, ref a..] if p.starts_with('p') => {
+                  let idx = parse_bits(&p[1..], "u32", "address space index")? as usize;
+                  let size = size(s, p)?;
+                  let align = align(a, p)?;
+                  resize_and_set(&mut dl.pointers, idx, Some((size, align)));
+                },
```

Comment:
Would we lose anything by enforcing that all address spaces have the same size & alignment for pointers?

Reply:
Yes, AMDGPU targets use non-uniform pointer sizes/alignments. Perhaps we don't need to keep the sizes/alignments around after this function returns, but I'd argue that it isn't worth the effort to remove it just for this PR: I'll just be adding it right back in and it's so trivial compared to the compiler as a whole.

Reply:
Why would you be adding it back in?

Reply:
Vega10 has a 64kB LDS (LDS is an explicitly managed CU (compute unit, conceptually similar to a thread/core) local cache) which has throughput in excess of 13TB/s (according to AMD documentation). Vs 484GB/s to Vega10's HDM2 main memory. Vega10 also has a cross CU GDS cache, but I don't have figures for its throughput. Of AMD Gpus, all GCN archs feature a LDS, and I'm reasonably certain Nvidia has something similar. Use of these memories is absolutely critical for maximizing GPGPU performance in memory throughput constrained code.

To use LDS, what is typically done is a global variable is declared and the input is copied into it at the start of a kernel, after which a barrier is executed before the real work of the kernel utilizing the LDS global is performed.

On `amdgcn-amd-amdhsa-amdgiz`, LDS addresses are 32bit (and 32bit aligned), not 64bit. GDS isn't technically implemented for HSA (not sure why, but I haven't dug into it), but would also be 32bit. We'll need access to this size/alignment info if we ever want to be able to support 32bit (vs uniform but double in size 64bit) references to data in this space.